stateresolution.go

```diff
@@ -0,0 +1,255 @@
+package gomatrixserverlib
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"sort"
+)
+
+// ResolveStateConflicts takes a list of state events with conflicting state keys
+// and works out which event should be used for each state event.
+func ResolveStateConflicts(conflicted []Event, authEvents []Event) []Event {
+	var r stateResolver
+	// Group the conflicted events by type and state key.
+	r.addConflicted(conflicted)
+	// Add the unconflicted auth events needed for auth checks.
+	for i := range authEvents {
+		r.addAuthEvent(&authEvents[i])
+	}
+	// Resolve the conflicted auth events.
+	r.resolveAndAddAuthBlocks([][]Event{r.creates})
+	r.resolveAndAddAuthBlocks([][]Event{r.powerLevels})
+	r.resolveAndAddAuthBlocks([][]Event{r.joinRules})
+	r.resolveAndAddAuthBlocks(r.thirdPartyInvites)
+	r.resolveAndAddAuthBlocks(r.members)
+	// Resolve any other conflicted state events.
+	for _, block := range r.others {
+		if event := r.resolveNormalBlock(block); event != nil {
+			r.result = append(r.result, *event)
+		}
+	}
+	return r.result
+}
+
+// A stateResolver tracks the internal state of the state resolution algorithm
+// It has 3 sections:
+//
+//  * Lists of lists of events to resolve grouped by event type and state key.
+//  * The resolved auth events grouped by type and state key.
+//  * A List of resolved events.
+//
+// It implements the AuthEvents interface and can be used for running auth checks.
+type stateResolver struct {
+	creates                   []Event
+	powerLevels               []Event
+	joinRules                 []Event
+	thirdPartyInvites         [][]Event
+	members                   [][]Event
+	others                    [][]Event
+	resolvedCreate            *Event
+	resolvedPowerLevels       *Event
+	resolvedJoinRules         *Event
+	resolvedThirdPartyInvites map[string]*Event
+	resolvedMembers           map[string]*Event
+	result                    []Event
```

Comment:
Are there any kind of guarantees made as to what `result` will contain? The API exposed by `stateResolver` implies the caller knows how this works and when `result` is nonsensical: I'm pretty sure you need to call `resolveAndAddAuthBlocks` for each event type for this to actually have the right result, and if you call `resolveAndAddAuthBlocks` too many times it'll also be incorrect.